### PR TITLE
BUG: Series.replace does not preserve dtype of original Series

### LINKED ISSUE #33484: BUG: pd.Series.replace(to_replace=...) does not preserve the original dtype of the series
- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the latest version of pandas.

- [X] (optional) I have confirmed this bug exists on the master branch of pandas.

---
#### Code Sample, a copy-pastable example

```python
>>> s = pd.Series(["one", "two"], dtype="string")
>>> expected = pd.Series(["1", "2"], dtype="string")
>>> result = s.replace(to_replace={"one": "1", "two": "2"})
>>> expected
0    1
1    2
dtype: string
>>> result
0    1
1    2
dtype: object
```
#### Problem description

The `pandas.Series.replace` does not return a series with the same `dtype` as the the original series. This behaviour is produced only when the replace function is called using the argument `to_replace`.

Notice: Similar behaviour is produced with other dtypes as well, this issue is not limited to string dtype.

The investigation should probably start here: `/pandas/core/generic.py`

#### Expected Output
We would expect the `pandas.Series.replace(to_replace=...)` to preserve the dtype of the series.
In the above code sample, both of the series should have a dtype of string.
#### Output of ``pd.show_versions()``

<details>
<h2>INSTALLED VERSIONS</h2>
commit           : None
python           : 3.8.2.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 19.4.0
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : None.UTF-8

pandas           : 1.0.3
numpy            : 1.18.1
pytz             : 2019.3
dateutil         : 2.8.1
pip              : 20.0.2
setuptools       : 46.1.3.post20200330
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : None
matplotlib       : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
pytest           : None
pyxlsb           : None
s3fs             : None
scipy            : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
xlsxwriter       : None
numba            : None
</details>

### PR DESCRIPTION
**Pull request by arw2019**
- [x] closes #33484
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Reheating #33622 (stale)

### CODE REVIEW

File: pandas/core/internals/blocks.py [hunk 1]
```diff
@@ -1174,6 +1177,9 @@ def coerce_to_target_dtype(self, other):
                 f"possible recursion in coerce_to_target_dtype: {self} {other}"
             )
 
+        if is_categorical_dtype(dtype) or self.is_datetime:
```
Comment (jreback) [inline pandas/core/internals/blocks.py:1180]: the is_datetime condition is already handled above, if for some reason you actually need to do this then do it above (and I am not really sure why you are trying to do this)

File: pandas/core/internals/blocks.py [hunk 2]
```diff
@@ -1130,7 +1130,10 @@ def coerce_to_target_dtype(self, other):
         if is_dtype_equal(self.dtype, dtype):
             return self
 
-        if self.is_bool or is_object_dtype(dtype) or is_bool_dtype(dtype):
+        if is_extension_array_dtype(self.dtype) and not is_categorical_dtype(dtype):
```
Comment (jreback) [inline pandas/core/internals/blocks.py:1133]: this if/then is already very complicated, what is the intent here?

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by arw2019** [comment]
Coverting to draft while I debug

**Comment by jbrockmendel** [comment]
@arw2019 the underlying problem (or one of them at least) is that ExtensionBlock._can_hold_element and ExtensionBlock.putmask are completely FUBAR.

I suggest you try adding the two following methods to CategoricalBlock (taken from a branch I have trying to get rid of CategoricalBlock.replace and _replace_list)

```
    def _can_hold_element(self, element: Any) -> bool:
        try:
            self.values._validate_setitem_value(element)
            return True
        except (TypeError, ValueError):
            return False

    def putmask(
        self, mask, new, inplace: bool = False, axis: int = 0, transpose: bool = False
    ) -> List["Block"]:
        if self._can_hold_element(new):
            return super().putmask(mask, new, inplace, axis, transpose)

        # TODO: should this be inplace?
        # TODO: use coerce_to_target_dtype?
        cat = self.values.add_categories(new)
        nb = self.make_block(cat)
        assert nb._can_hold_element(new)
        return nb.putmask(mask, new, inplace, axis, transpose)
```

and in coerce_to_target_dtype just after the is_dtype_equal(self.dtype, dtype) check:

```
        if self._can_hold_element(other):
            return self
        elif self.is_categorical:
            # Note: this will be wrong if we ever have a tuple category
            cat = self.values.add_categories(other)
            return self.make_block(cat)
```

This changes the behavior for a handful of tests causing Series/DataFrame.replace to return CategoricalDtype (with added categories) instead of object-dtype, so may not be _quite_ what we want.  But I'm pretty sure its in the right ballpark.

**Comment by arw2019** [comment]
> if self._can_hold_element(other):
>             return self
>         elif self.is_categorical:
>             # Note: this will be wrong if we ever have a tuple category
>             cat = self.values.add_categories(other)
>             return self.make_block(cat)

Thanks!!! Pushed this, will see what fails

**Comment by jbrockmendel** [comment]
cc @jorisvandenbossche the latest commit here is pretty much exactly the edits i was referring to re CategoricalBlock.replace.  Only difference is I would also remove CategoricalBlock.replace and _replace_list.

**Pull request closed without merging**